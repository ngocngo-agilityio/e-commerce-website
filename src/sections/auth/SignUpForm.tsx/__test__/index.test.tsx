import { fireEvent, render, waitFor, screen } from '@testing-library/react';

// Sections
import SignUpSection from '..';

// Mocks
import {
  MOCK_ERROR_MESSAGES,
  MOCK_SIGN_IN_PAYLOAD,
  MOCK_SIGN_UP_PAYLOAD,
} from '@mocks';

// Constants
import { APP_ROUTERS, SUCCESS_MESSAGES } from '@constants';

const mockPush = jest.fn();
const mockSignUp = jest.fn();
const mockCreateCart = jest.fn();

jest.mock('next/navigation', () => ({
  ...jest.requireActual('next/navigation'),
  useRouter: () => ({
    push: mockPush,
  }),
}));

jest.mock('@actions', () => ({
  ...jest.requireActual('@actions'),
  signUp: () => mockSignUp(),
  createCart: () => mockCreateCart(),
}));

describe('SignUpSection', () => {
  const submitForm = () => {
    const FIRST_NAME = screen.getByPlaceholderText('First name');
    const LAST_NAME = screen.getByPlaceholderText('Last name');
    const EMAIL_FIELD = screen.getByPlaceholderText('Email');
    const PASSWORD_FIELD = screen.getByPlaceholderText('Password');
    const CONFIRM_PASSWORD_FIELD =
      screen.getByPlaceholderText('Confirm password');
    const SUBMIT_BTN = screen.getByText('Sign up');

    // Change First name field
    fireEvent.change(FIRST_NAME, {
      target: { value: MOCK_SIGN_UP_PAYLOAD.firstName },
    });

    // Change Last name field
    fireEvent.change(LAST_NAME, {
      target: { value: MOCK_SIGN_UP_PAYLOAD.lastName },
    });

    // Change Email field
    fireEvent.change(EMAIL_FIELD, {
      target: { value: MOCK_SIGN_UP_PAYLOAD.email },
    });

    // Change Password field
    fireEvent.change(PASSWORD_FIELD, {
      target: { value: MOCK_SIGN_UP_PAYLOAD.password },
    });

    // Change Confirm Password field
    fireEvent.change(CONFIRM_PASSWORD_FIELD, {
      target: { value: MOCK_SIGN_UP_PAYLOAD.password },
    });

    fireEvent.click(SUBMIT_BTN);
  };

  test('should match with snapshot', () => {
    const { container } = render(<SignUpSection />);

    expect(container).toMatchSnapshot();
  });

  test('should sign up successfully', () => {
    mockSignUp.mockReturnValue(null);
    mockCreateCart.mockReturnValue(undefined);
    const { getByText } = render(<SignUpSection />);

    submitForm();

    waitFor(() => {
      expect(mockSignUp).toHaveBeenCalledWith(MOCK_SIGN_IN_PAYLOAD);
      expect(getByText(SUCCESS_MESSAGES.SIGN_UP)).toBeInTheDocument();
      expect(mockPush).toHaveBeenCalledWith(APP_ROUTERS.SIGN_IN);
    });
  });

  test('should sign up failed with error while call api sign up', () => {
    mockSignUp.mockReturnValue({ error: MOCK_ERROR_MESSAGES });
    mockCreateCart.mockReturnValue({ error: MOCK_ERROR_MESSAGES });
    const { getByText } = render(<SignUpSection />);

    submitForm();

    waitFor(() => {
      expect(mockSignUp).toHaveBeenCalledWith(MOCK_SIGN_IN_PAYLOAD);
      expect(getByText(MOCK_ERROR_MESSAGES)).toBeInTheDocument();
    });
  });

  test('should sign up failed with error while call api to create cart', () => {
    mockSignUp.mockReturnValue(null);
    mockCreateCart.mockReturnValue({ error: MOCK_ERROR_MESSAGES });
    const { getByText } = render(<SignUpSection />);

    submitForm();

    waitFor(() => {
      expect(mockSignUp).toHaveBeenCalledWith(MOCK_SIGN_IN_PAYLOAD);
      expect(getByText(MOCK_ERROR_MESSAGES)).toBeInTheDocument();
    });
  });
});
