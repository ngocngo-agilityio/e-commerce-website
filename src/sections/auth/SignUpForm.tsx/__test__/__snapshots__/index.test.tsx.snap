// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpSection should match with snapshot 1`] = `
<div>
  <form
    class="css-1m7lw4y"
  >
    <div
      class="css-4slh5a"
    >
      <div
        class="css-b95f0i"
      >
        <label
          class="chakra-form__label css-uvho5g"
        >
          First Name
        </label>
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <input
            class="chakra-input css-0"
            id="field-:r2:"
            name="firstName"
            placeholder="First name"
            value=""
          />
        </div>
      </div>
      <div
        class="css-b95f0i"
      >
        <label
          class="chakra-form__label css-uvho5g"
        >
          Last Name
        </label>
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <input
            class="chakra-input css-0"
            id="field-:r3:"
            name="lastName"
            placeholder="Last name"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-j7qwjs"
    >
      <label
        class="chakra-form__label css-uvho5g"
      >
        Email
      </label>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <input
          class="chakra-input css-0"
          id="field-:r4:"
          name="email"
          placeholder="Email"
          value=""
        />
      </div>
    </div>
    <div
      class="css-j7qwjs"
    >
      <label
        class="chakra-form__label css-uvho5g"
      >
        Password
      </label>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            class="chakra-input css-0"
            id="field-:r5:"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <div
            class="chakra-input__right-element css-dp9r5f"
            data-testid="toggle-password-icon"
            style="cursor: pointer;"
          >
            <svg
              class="chakra-icon css-w31a5m"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g
                fill="currentColor"
              >
                <path
                  d="M23.432,10.524C20.787,7.614,16.4,4.538,12,4.6,7.6,4.537,3.213,7.615.568,10.524a2.211,2.211,0,0,0,0,2.948C3.182,16.351,7.507,19.4,11.839,19.4h.308c4.347,0,8.671-3.049,11.288-5.929A2.21,2.21,0,0,0,23.432,10.524ZM7.4,12A4.6,4.6,0,1,1,12,16.6,4.6,4.6,0,0,1,7.4,12Z"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="2"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-j7qwjs"
    >
      <label
        class="chakra-form__label css-uvho5g"
      >
        Confirm Password
      </label>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            class="chakra-input css-0"
            id="field-:r6:"
            name="confirmPassword"
            placeholder="Confirm password"
            type="password"
            value=""
          />
          <div
            class="chakra-input__right-element css-dp9r5f"
            data-testid="toggle-confirm-password-icon"
            style="cursor: pointer;"
          >
            <svg
              class="chakra-icon css-w31a5m"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g
                fill="currentColor"
              >
                <path
                  d="M23.432,10.524C20.787,7.614,16.4,4.538,12,4.6,7.6,4.537,3.213,7.615.568,10.524a2.211,2.211,0,0,0,0,2.948C3.182,16.351,7.507,19.4,11.839,19.4h.308c4.347,0,8.671-3.049,11.288-5.929A2.21,2.21,0,0,0,23.432,10.524ZM7.4,12A4.6,4.6,0,1,1,12,16.6,4.6,4.6,0,0,1,7.4,12Z"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="2"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <button
      class="chakra-button css-1kz4kx3"
      disabled=""
      type="submit"
    >
      Sign up
    </button>
  </form>
</div>
`;
